File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.cc
#include <cstdio>
#include <queue>
#include <vector>
#define inf 0x7fffffff

using std::priority_queue;
using std::vector;

int n, m, dist[1005];

struct edge_t {
    int node, weight;

    edge_t(int _node, int _weight) : node(_node), weight(_weight) {}

    bool operator< (const edge_t & target) const {
        return this->weight > target.weight;
    }
};

vector<edge_t> arr[1005];

void dijkstra(int start) {
    for(int i = 1; i <= n; ++i) dist[i] = inf;
    dist[start] = 0;

    priority_queue<edge_t> pq;
    pq.push(edge_t(start, 0));

    while(!pq.empty()) {
        edge_t curr = pq.top(); pq.pop();
        
        if(dist[curr.node] < curr.weight) continue;

        for(const edge_t &x : arr[curr.node]) {
            if(dist[x.node] <= dist[curr.node] + x.weight) continue;

            dist[x.node] = dist[curr.node] + x.weight;
            pq.push(edge_t(x.node, dist[x.node]));
        }
    }
}

int main() {
    // input
    scanf("%d\n%d", &n, &m);
    
    for(int i = 0; i < m; ++i) {
        int x, y, w; scanf("%d %d %d", &x, &y,&w);
        arr[x].push_back(edge_t(y, w));
    }

    int start, end; scanf("%d\n%d", &start, &end);

    // dijkstra
    dijkstra(start);

    // output
    printf("%d", dist[end]);
    
    return 0;
}

File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.py
import sys
from heapq import heappush, heappop
si = sys.stdin.readline

n = int(si())
m = int(si())

arr = [[] for _ in range(n+1)]
dist = [float("inf")] * (n + 1)

class Edge:
    def __init__(self, node:int, weight:int):
        self.node = node
        self.weight = weight

    def __lt__(self, other):
        return self.weight < other.weight
    
    def __iter__(self):
        return iter((self.node, self.weight))

    def __str__(self):
        return f"<{self.node}, {self.weight}>"
    
for _ in range(m):
    x, y, w = map(int, si().split())
    arr[x].append(Edge(y,w))

start, end = map(int, si().split())
# dijkstra
pq = []
dist[start] = 0
heappush(pq, Edge(start, 0))

while pq:
    node, weight = heappop(pq)

    if dist[node] < weight: continue

    for n, e in arr[node]:
        if dist[n] <= dist[node] + e: continue

        dist[n] = dist[node] + e
        heappush(pq, Edge(n, dist[n]))

print(dist[end])
